### PR TITLE
refactor: implement new qp2p version

### PR DESCRIPTION
**Pull request by b-zee**
This introduces quite some changes to the API that hopefully simplifies
much of the internals of the node and client.

### CODE REVIEW

**Comment by b-zee** [comment]
bors try

**Comment by b-zee** [comment]
bors try

**Comment by b-zee** [comment]
bors try

File: sn_comms/src/link.rs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+            let _handle = tokio::spawn(async move {
+                // Attempt to gracefully terminate the stream.
+                // If this errors it does _not_ mean our message has not been sent
+                let result = send_stream.finish().await;
+                trace!("{msg_id:?} finished {stream_id_clone} to {peer:?}: {result:?}");
+            });
+
+            match recv_stream.read().await {
+                Ok(response) => break Ok(response),
+                // Ok(None) => {
```
Comment (joshuef) [inline sn_comms/src/link.rs:174]: lets clear up this

File: sn_node/src/node/cfg/config_handler.rs [hunk 2]
```diff
@@ -170,31 +139,31 @@ impl Config {
...
-        if local_ip_unspecified && self.first && self.public_addr.is_none() {
-            return Err(Error::Configuration(
-                "Must specify public address for --first node. \
-                The first node cannot query its public address from peers, so one must be \
-                specifed. This can be specified with --public-addr, or by setting a concrete IP \
-                for --local-addr."
-                    .to_string(),
-            ));
-        }
+        // if let Some(local_addr) = self.local_addr {
```
Comment (joshuef) [inline sn_node/src/node/cfg/config_handler.rs:142]: more to cleanup

**Review by joshuef — CHANGES_REQUESTED** [review]

**Review by joshuef — APPROVED** [review]
bors r+

**Pull request merged**